More verbose ssh failure error messages.

alikins opened the pull request:
##### ISSUE TYPE

<!--- Pick one below and delete the rest: -->
- Bugfix Pull Request
##### COMPONENT NAME

<!--- Name of the plugin/module/task -->

lib/ansible/plugins/connection/ssh.py
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
ansible 2.2.0 (minimal_better_ssh_failures 6b5627ad93) last updated 2016/09/15 17:05:00 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 48d932643b) last updated 2016/09/15 17:01:15 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD aa45bd8a94) last updated 2016/09/15 17:01:15 (GMT -400)
  config file = /home/adrian/ansible/ansible.cfg
  configured module search path = ['/home/adrian/src/ansible-modules-core', '/home/adrian/src/ansible-modules-extras']

```
##### SUMMARY

Don't force 'ssh -q' for verbosity < 'vvv'.

Include ssh stderr in exception **str** for
ssh connection errors.

Fixes #15706

alikins commented:
Before:

```
[fedora-23:ansible (devel % u=)]$ ansible-playbook ping.yml 

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
fatal: [docker]: UNREACHABLE! => {"changed": false, "msg": "Authentication or permission failure. In some cases, you may have been able to authenticate and did not have permissions on the remote directory. Consider changing the remote temp path in ansible.cfg to a path rooted in \"/tmp\". Failed command was: ( umask 77 && mkdir -p \"` echo $HOME/.ansible/tmp/ansible-tmp-1473973687.6-45437967573517 `\" && echo ansible-tmp-1473973687.6-45437967573517=\"` echo $HOME/.ansible/tmp/ansible-tmp-1473973687.6-45437967573517 `\" ), exited with result 1", "unreachable": true}
fatal: [noname]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}
fatal: [grimlock]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}
fatal: [cantconnect]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}
fatal: [vagrant-ubuntu]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}
The authenticity of host 'bastion (127.0.0.1)' can't be established.
ECDSA key fingerprint is SHA256:AnDbPYQqQemno0nUxK4UpSUxUDhXncimH0b3j864mtE.
ECDSA key fingerprint is MD5:ff:3d:00:10:7c:0a:97:d2:75:6e:0c:96:56:ec:45:9f.
Are you sure you want to continue connecting (yes/no)? ok: [me]
ok: [localhost]
ok: [swoop]
fatal: [test@swoop]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}
fatal: [adrian@swoop]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}
fatal: [adrian@me]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}
fatal: [default]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}

TASK [ping] ********************************************************************
ok: [me]
ok: [localhost]
ok: [swoop]
    to retry, use: --limit @ping.retry

PLAY RECAP *********************************************************************
adrian@me                  : ok=0    changed=0    unreachable=1    failed=0   
adrian@swoop               : ok=0    changed=0    unreachable=1    failed=0   
cantconnect                : ok=0    changed=0    unreachable=1    failed=0   
default                    : ok=0    changed=0    unreachable=1    failed=0   
docker                     : ok=0    changed=0    unreachable=1    failed=0   
grimlock                   : ok=0    changed=0    unreachable=1    failed=0   
localhost                  : ok=2    changed=0    unreachable=0    failed=0   
me                         : ok=2    changed=0    unreachable=0    failed=0   
noname                     : ok=0    changed=0    unreachable=1    failed=0   
swoop                      : ok=2    changed=0    unreachable=0    failed=0   
test@swoop                 : ok=0    changed=0    unreachable=1    failed=0   
vagrant-ubuntu             : ok=0    changed=0    unreachable=1    failed=0   

```

After:

```
[fedora-23:ansible (minimal_better_ssh_failures %)]$ ansible-playbook ping.yml 

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
fatal: [docker]: UNREACHABLE! => {"changed": false, "msg": "Authentication or permission failure. In some cases, you may have been able to authenticate and did not have permissions on the remote directory. Consider changing the remote temp path in ansible.cfg to a path rooted in \"/tmp\". Failed command was: ( umask 77 && mkdir -p \"` echo $HOME/.ansible/tmp/ansible-tmp-1473973715.96-210220887784299 `\" && echo ansible-tmp-1473973715.96-210220887784299=\"` echo $HOME/.ansible/tmp/ansible-tmp-1473973715.96-210220887784299 `\" ), exited with result 1", "unreachable": true}
fatal: [noname]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.\nstderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016\r\ndebug1: Reading configuration data /home/adrian/.ssh/config\r\ndebug1: /home/adrian/.ssh/config line 62: Applying options for *\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 56: Applying options for *\r\ndebug1: /etc/ssh/ssh_config line 67: Deprecated option \"useroaming\"\r\ndebug2: resolving \"noname.g.a\" port 22\r\nssh: Could not resolve hostname noname.g.a: No address associated with hostname\r\n"}
fatal: [grimlock]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.\nstderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016\r\ndebug1: Reading configuration data /home/adrian/.ssh/config\r\ndebug1: /home/adrian/.ssh/config line 20: Applying options for 192.168.1.36\r\ndebug1: /home/adrian/.ssh/config line 41: Applying options for 192.168.*\r\ndebug1: /home/adrian/.ssh/config line 62: Applying options for *\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 56: Applying options for *\r\ndebug1: /etc/ssh/ssh_config line 67: Deprecated option \"useroaming\"\r\ndebug2: resolving \"192.168.1.36\" port 22\r\ndebug2: ssh_connect_direct: needpriv 0\r\ndebug1: Connecting to 192.168.1.36 [192.168.1.36] port 22.\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug1: connect to address 192.168.1.36 port 22: No route to host\r\nssh: connect to host 192.168.1.36 port 22: No route to host\r\n"}
ok: [swoop]
The authenticity of host 'bastion (127.0.0.1)' can't be established.
ECDSA key fingerprint is SHA256:AnDbPYQqQemno0nUxK4UpSUxUDhXncimH0b3j864mtE.
ECDSA key fingerprint is MD5:ff:3d:00:10:7c:0a:97:d2:75:6e:0c:96:56:ec:45:9f.
Are you sure you want to continue connecting (yes/no)? fatal: [vagrant-ubuntu]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.\nstderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016\r\ndebug1: Reading configuration data /home/adrian/.ssh/config\r\ndebug1: /home/adrian/.ssh/config line 41: Applying options for 192.168.*\r\ndebug1: /home/adrian/.ssh/config line 62: Applying options for *\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 56: Applying options for *\r\ndebug1: /etc/ssh/ssh_config line 67: Deprecated option \"useroaming\"\r\ndebug2: resolving \"192.168.121.231\" port 22\r\ndebug2: ssh_connect_direct: needpriv 0\r\ndebug1: Connecting to 192.168.121.231 [192.168.121.231] port 22.\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug1: connect to address 192.168.121.231 port 22: No route to host\r\nssh: connect to host 192.168.121.231 port 22: No route to host\r\n"}
fatal: [cantconnect]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.\nstderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016\r\ndebug1: Reading configuration data /home/adrian/.ssh/config\r\ndebug1: /home/adrian/.ssh/config line 41: Applying options for 192.168.*\r\ndebug1: /home/adrian/.ssh/config line 62: Applying options for *\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 56: Applying options for *\r\ndebug1: /etc/ssh/ssh_config line 67: Deprecated option \"useroaming\"\r\ndebug2: resolving \"192.168.1.253\" port 22\r\ndebug2: ssh_connect_direct: needpriv 0\r\ndebug1: Connecting to 192.168.1.253 [192.168.1.253] port 22.\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug1: connect to address 192.168.1.253 port 22: No route to host\r\nssh: connect to host 192.168.1.253 port 22: No route to host\r\n"}
ok: [localhost]
ok: [me]
fatal: [test@swoop]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.\nstderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016\r\ndebug1: Reading configuration data /home/adrian/.ssh/config\r\ndebug1: /home/adrian/.ssh/config line 62: Applying options for *\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 56: Applying options for *\r\ndebug1: /etc/ssh/ssh_config line 67: Deprecated option \"useroaming\"\r\ndebug2: resolving \"182.168.1.120\" port 22\r\ndebug2: ssh_connect_direct: needpriv 0\r\ndebug1: Connecting to 182.168.1.120 [182.168.1.120] port 22.\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug1: connect to address 182.168.1.120 port 22: Connection timed out\r\nssh: connect to host 182.168.1.120 port 22: Connection timed out\r\n"}
fatal: [adrian@swoop]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.\nstderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016\r\ndebug1: Reading configuration data /home/adrian/.ssh/config\r\ndebug1: /home/adrian/.ssh/config line 62: Applying options for *\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 56: Applying options for *\r\ndebug1: /etc/ssh/ssh_config line 67: Deprecated option \"useroaming\"\r\ndebug2: resolving \"182.168.1.120\" port 22\r\ndebug2: ssh_connect_direct: needpriv 0\r\ndebug1: Connecting to 182.168.1.120 [182.168.1.120] port 22.\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug1: connect to address 182.168.1.120 port 22: Connection timed out\r\nssh: connect to host 182.168.1.120 port 22: Connection timed out\r\n"}
fatal: [adrian@me]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.\nstderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016\r\ndebug1: Reading configuration data /home/adrian/.ssh/config\r\ndebug1: /home/adrian/.ssh/config line 14: Applying options for thinkpad-wired.g.a\r\ndebug1: /home/adrian/.ssh/config line 62: Applying options for *\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 56: Applying options for *\r\ndebug1: /etc/ssh/ssh_config line 67: Deprecated option \"useroaming\"\r\ndebug1: Executing proxy command: exec ssh adrian@bastion -W thinkpad-wired.g.a:22\r\ndebug3: timeout: 10000 ms remain after connect\r\ndebug1: permanently_drop_suid: 1000\r\ndebug1: identity file /home/adrian/.ssh/id_rsa type 1\r\ndebug1: key_load_public: No such file or directory\r\ndebug1: identity file /home/adrian/.ssh/id_rsa-cert type -1\r\ndebug1: key_load_public: No such file or directory\r\ndebug1: identity file /home/adrian/.ssh/id_dsa type -1\r\ndebug1: key_load_public: No such file or directory\r\ndebug1: identity file /home/adrian/.ssh/id_dsa-cert type -1\r\ndebug1: key_load_public: No such file or directory\r\ndebug1: identity file /home/adrian/.ssh/id_ecdsa type -1\r\ndebug1: key_load_public: No such file or directory\r\ndebug1: identity file /home/adrian/.ssh/id_ecdsa-cert type -1\r\ndebug1: key_load_public: No such file or directory\r\ndebug1: identity file /home/adrian/.ssh/id_ed25519 type -1\r\ndebug1: key_load_public: No such file or directory\r\ndebug1: identity file /home/adrian/.ssh/id_ed25519-cert type -1\r\ndebug1: Enabling compatibility mode for protocol 2.0\r\ndebug1: Local version string SSH-2.0-OpenSSH_7.2\r\nConnection timed out during banner exchange\r\n"}
fatal: [default]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.\nstderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016\r\ndebug1: Reading configuration data /home/adrian/.ssh/config\r\ndebug1: /home/adrian/.ssh/config line 62: Applying options for *\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 56: Applying options for *\r\ndebug1: /etc/ssh/ssh_config line 67: Deprecated option \"useroaming\"\r\ndebug2: resolving \"default\" port 22\r\ndebug2: ssh_connect_direct: needpriv 0\r\ndebug1: Connecting to default [10.0.2.1] port 22.\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug1: connect to address 10.0.2.1 port 22: Connection timed out\r\nssh: connect to host default port 22: Connection timed out\r\n"}

TASK [ping] ********************************************************************
ok: [localhost]
ok: [me]
ok: [swoop]
    to retry, use: --limit @ping.retry

PLAY RECAP *********************************************************************
adrian@me                  : ok=0    changed=0    unreachable=0    failed=1   
adrian@swoop               : ok=0    changed=0    unreachable=0    failed=1   
cantconnect                : ok=0    changed=0    unreachable=0    failed=1   
default                    : ok=0    changed=0    unreachable=0    failed=1   
docker                     : ok=0    changed=0    unreachable=1    failed=0   
grimlock                   : ok=0    changed=0    unreachable=0    failed=1   
localhost                  : ok=2    changed=0    unreachable=0    failed=0   
me                         : ok=2    changed=0    unreachable=0    failed=0   
noname                     : ok=0    changed=0    unreachable=0    failed=1   
swoop                      : ok=2    changed=0    unreachable=0    failed=0   
test@swoop                 : ok=0    changed=0    unreachable=0    failed=1   
vagrant-ubuntu             : ok=0    changed=0    unreachable=0    failed=1   

```

alikins commented:
And better, with error display unescaping strings:

```
[fedora-23:ansible (minimal_better_ssh_failures %)]$ ansible-playbook ping.yml 

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
fatal: [docker]: UNREACHABLE! => {"changed": false, "msg": "Authentication or permission failure. In some cases, you may have been able to authenticate and did not have permissions on the remote directory. Consider changing the remote temp path in ansible.cfg to a path rooted in "/tmp". Failed command was: ( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1473974050.65-268256174954466 `" && echo ansible-tmp-1473974050.65-268256174954466="` echo $HOME/.ansible/tmp/ansible-tmp-1473974050.65-268256174954466 `" ), exited with result 1", "unreachable": true}
fatal: [noname]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016
debug1: Reading configuration data /home/adrian/.ssh/config
debug1: /home/adrian/.ssh/config line 62: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 56: Applying options for *
debug1: /etc/ssh/ssh_config line 67: Deprecated option "useroaming"
debug2: resolving "noname.g.a" port 22
ssh: Could not resolve hostname noname.g.a: No address associated with hostname
"}
ok: [swoop]
fatal: [grimlock]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016
debug1: Reading configuration data /home/adrian/.ssh/config
debug1: /home/adrian/.ssh/config line 20: Applying options for 192.168.1.36
debug1: /home/adrian/.ssh/config line 41: Applying options for 192.168.*
debug1: /home/adrian/.ssh/config line 62: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 56: Applying options for *
debug1: /etc/ssh/ssh_config line 67: Deprecated option "useroaming"
debug2: resolving "192.168.1.36" port 22
debug2: ssh_connect_direct: needpriv 0
debug1: Connecting to 192.168.1.36 [192.168.1.36] port 22.
debug2: fd 3 setting O_NONBLOCK
debug1: connect to address 192.168.1.36 port 22: No route to host
ssh: connect to host 192.168.1.36 port 22: No route to host
"}
The authenticity of host 'bastion (127.0.0.1)' can't be established.
ECDSA key fingerprint is SHA256:AnDbPYQqQemno0nUxK4UpSUxUDhXncimH0b3j864mtE.
ECDSA key fingerprint is MD5:ff:3d:00:10:7c:0a:97:d2:75:6e:0c:96:56:ec:45:9f.
Are you sure you want to continue connecting (yes/no)? fatal: [vagrant-ubuntu]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016
debug1: Reading configuration data /home/adrian/.ssh/config
debug1: /home/adrian/.ssh/config line 41: Applying options for 192.168.*
debug1: /home/adrian/.ssh/config line 62: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 56: Applying options for *
debug1: /etc/ssh/ssh_config line 67: Deprecated option "useroaming"
debug2: resolving "192.168.121.231" port 22
debug2: ssh_connect_direct: needpriv 0
debug1: Connecting to 192.168.121.231 [192.168.121.231] port 22.
debug2: fd 3 setting O_NONBLOCK
debug1: connect to address 192.168.121.231 port 22: No route to host
ssh: connect to host 192.168.121.231 port 22: No route to host
"}
fatal: [cantconnect]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016
debug1: Reading configuration data /home/adrian/.ssh/config
debug1: /home/adrian/.ssh/config line 41: Applying options for 192.168.*
debug1: /home/adrian/.ssh/config line 62: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 56: Applying options for *
debug1: /etc/ssh/ssh_config line 67: Deprecated option "useroaming"
debug2: resolving "192.168.1.253" port 22
debug2: ssh_connect_direct: needpriv 0
debug1: Connecting to 192.168.1.253 [192.168.1.253] port 22.
debug2: fd 3 setting O_NONBLOCK
debug1: connect to address 192.168.1.253 port 22: No route to host
ssh: connect to host 192.168.1.253 port 22: No route to host
"}
ok: [localhost]
ok: [me]
fatal: [test@swoop]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016
debug1: Reading configuration data /home/adrian/.ssh/config
debug1: /home/adrian/.ssh/config line 62: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 56: Applying options for *
debug1: /etc/ssh/ssh_config line 67: Deprecated option "useroaming"
debug2: resolving "182.168.1.120" port 22
debug2: ssh_connect_direct: needpriv 0
debug1: Connecting to 182.168.1.120 [182.168.1.120] port 22.
debug2: fd 3 setting O_NONBLOCK
debug1: connect to address 182.168.1.120 port 22: Connection timed out
ssh: connect to host 182.168.1.120 port 22: Connection timed out
"}
fatal: [adrian@swoop]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016
debug1: Reading configuration data /home/adrian/.ssh/config
debug1: /home/adrian/.ssh/config line 62: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 56: Applying options for *
debug1: /etc/ssh/ssh_config line 67: Deprecated option "useroaming"
debug2: resolving "182.168.1.120" port 22
debug2: ssh_connect_direct: needpriv 0
debug1: Connecting to 182.168.1.120 [182.168.1.120] port 22.
debug2: fd 3 setting O_NONBLOCK
debug1: connect to address 182.168.1.120 port 22: Connection timed out
ssh: connect to host 182.168.1.120 port 22: Connection timed out
"}
fatal: [adrian@me]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016
debug1: Reading configuration data /home/adrian/.ssh/config
debug1: /home/adrian/.ssh/config line 14: Applying options for thinkpad-wired.g.a
debug1: /home/adrian/.ssh/config line 62: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 56: Applying options for *
debug1: /etc/ssh/ssh_config line 67: Deprecated option "useroaming"
debug1: Executing proxy command: exec ssh adrian@bastion -W thinkpad-wired.g.a:22
debug3: timeout: 10000 ms remain after connect
debug1: permanently_drop_suid: 1000
debug1: identity file /home/adrian/.ssh/id_rsa type 1
debug1: key_load_public: No such file or directory
debug1: identity file /home/adrian/.ssh/id_rsa-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file /home/adrian/.ssh/id_dsa type -1
debug1: key_load_public: No such file or directory
debug1: identity file /home/adrian/.ssh/id_dsa-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file /home/adrian/.ssh/id_ecdsa type -1
debug1: key_load_public: No such file or directory
debug1: identity file /home/adrian/.ssh/id_ecdsa-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file /home/adrian/.ssh/id_ed25519 type -1
debug1: key_load_public: No such file or directory
debug1: identity file /home/adrian/.ssh/id_ed25519-cert type -1
debug1: Enabling compatibility mode for protocol 2.0
debug1: Local version string SSH-2.0-OpenSSH_7.2
Connection timed out during banner exchange
"}
fatal: [default]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: OpenSSH_7.2p2, OpenSSL 1.0.2h-fips  3 May 2016
debug1: Reading configuration data /home/adrian/.ssh/config
debug1: /home/adrian/.ssh/config line 62: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 56: Applying options for *
debug1: /etc/ssh/ssh_config line 67: Deprecated option "useroaming"
debug2: resolving "default" port 22
debug2: ssh_connect_direct: needpriv 0
debug1: Connecting to default [10.0.2.1] port 22.
debug2: fd 3 setting O_NONBLOCK
debug1: connect to address 10.0.2.1 port 22: Connection timed out
ssh: connect to host default port 22: Connection timed out
"}

TASK [ping] ********************************************************************
ok: [localhost]
ok: [me]
ok: [swoop]
    to retry, use: --limit @ping.retry

PLAY RECAP *********************************************************************
adrian@me                  : ok=0    changed=0    unreachable=0    failed=1   
adrian@swoop               : ok=0    changed=0    unreachable=0    failed=1   
cantconnect                : ok=0    changed=0    unreachable=0    failed=1   
default                    : ok=0    changed=0    unreachable=0    failed=1   
docker                     : ok=0    changed=0    unreachable=1    failed=0   
grimlock                   : ok=0    changed=0    unreachable=0    failed=1   
localhost                  : ok=2    changed=0    unreachable=0    failed=0   
me                         : ok=2    changed=0    unreachable=0    failed=0   
noname                     : ok=0    changed=0    unreachable=0    failed=1   
swoop                      : ok=2    changed=0    unreachable=0    failed=0   
test@swoop                 : ok=0    changed=0    unreachable=0    failed=1   
vagrant-ubuntu             : ok=0    changed=0    unreachable=0    failed=1   

```

alikins commented:
Even better (without a local config of 'ssh_args = -vvv'):

```
[fedora-23:ansible (minimal_better_ssh_failures %)]$ ansible-playbook ping.yml 

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
fatal: [docker]: UNREACHABLE! => {"changed": false, "msg": "Authentication or permission failure. In some cases, you may have been able to authenticate and did not have permissions on the remote directory. Consider changing the remote temp path in ansible.cfg to a path rooted in "/tmp". Failed command was: ( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1473974152.79-165291990282070 `" && echo ansible-tmp-1473974152.79-165291990282070="` echo $HOME/.ansible/tmp/ansible-tmp-1473974152.79-165291990282070 `" ), exited with result 1", "unreachable": true}
fatal: [noname]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: ssh: Could not resolve hostname noname.g.a: No address associated with hostname
"}
fatal: [grimlock]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: ssh: connect to host 192.168.1.36 port 22: No route to host
"}
ok: [swoop]
The authenticity of host 'bastion (127.0.0.1)' can't be established.
ECDSA key fingerprint is SHA256:AnDbPYQqQemno0nUxK4UpSUxUDhXncimH0b3j864mtE.
ECDSA key fingerprint is MD5:ff:3d:00:10:7c:0a:97:d2:75:6e:0c:96:56:ec:45:9f.
Are you sure you want to continue connecting (yes/no)? fatal: [cantconnect]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: ssh: connect to host 192.168.1.253 port 22: No route to host
"}
fatal: [vagrant-ubuntu]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: ssh: connect to host 192.168.121.231 port 22: No route to host
"}
ok: [localhost]
ok: [me]
fatal: [test@swoop]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: ssh: connect to host 182.168.1.120 port 22: Connection timed out
"}
fatal: [adrian@swoop]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: ssh: connect to host 182.168.1.120 port 22: Connection timed out
"}
fatal: [adrian@me]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: Connection timed out during banner exchange
"}
fatal: [default]: FAILED! => {"failed": true, "msg": "Failed to connect to the host via ssh.
stderr: ssh: connect to host default port 22: Connection timed out
"}

TASK [ping] ********************************************************************
ok: [me]
ok: [localhost]
ok: [swoop]
    to retry, use: --limit @ping.retry

PLAY RECAP *********************************************************************
adrian@me                  : ok=0    changed=0    unreachable=0    failed=1   
adrian@swoop               : ok=0    changed=0    unreachable=0    failed=1   
cantconnect                : ok=0    changed=0    unreachable=0    failed=1   
default                    : ok=0    changed=0    unreachable=0    failed=1   
docker                     : ok=0    changed=0    unreachable=1    failed=0   
grimlock                   : ok=0    changed=0    unreachable=0    failed=1   
localhost                  : ok=2    changed=0    unreachable=0    failed=0   
me                         : ok=2    changed=0    unreachable=0    failed=0   
noname                     : ok=0    changed=0    unreachable=0    failed=1   
swoop                      : ok=2    changed=0    unreachable=0    failed=0   
test@swoop                 : ok=0    changed=0    unreachable=0    failed=1   
vagrant-ubuntu             : ok=0    changed=0    unreachable=0    failed=1   

```

alikins commented:
(The escape.decode() likely needs some tweaks to work generally, but my other more complicated ssh error pr seemed to do okay)

alikins committed:
More verbose ssh failure error messages.

Don't force 'ssh -q' for verbosity < 'vvv'.

Include ssh stderr in exception __str__ for
ssh connection errors.

Fixes #15706

bcoca commented on lib/ansible/errors/__init__.py:
```diff
@@ -173,6 +173,19 @@ class AnsibleConnectionFailure(AnsibleRuntimeError):
     ''' the transport / connection_plugin had a fatal error '''
     pass
 
+class AnsibleSshConnectionFailure(AnsibleRuntimeError):
```
it should be subclass of AnsibleConnectionFailure at least to be able to handle these genericly

alikins replied:
Fixed now.

bcoca commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -144,9 +144,9 @@ def _build_command(self, binary, *other_args):
-            self._command += ['-q']
+        #elif binary == self._play_context.ssh_executable:
+        #    # Older versions of ssh (e.g. in RHEL 6) don't accept sftp -q.
+        #    self._command += ['-q']
```
just remove, don't comment out

alikins replied:
removed now.

alikins commented:
For something that also has more readable output of the stderr from ssh (sans codecs.escape_decode()):

https://github.com/ansible/ansible/compare/devel...alikins:minimal_better_ssh_failures_no_escape?expand=1

alikins commented:
This would also fix https://github.com/ansible/ansible/issues/15525

bcoca commented on lib/ansible/executor/task_executor.py:
```diff
@@ -486,7 +486,10 @@ def _execute(self, variables=None):
-                return dict(unreachable=True, msg=to_text(e))
+                failure_dict = dict(unreachable=True,
+                                    msg=to_text(e),
+                                    _ansible_connection_exception=e)
```
do not pass an exception object into the result, this turns into JSON datastructures and we don't want to serialize/deserialize exceptions, specially since you use as a boolean on the other side.

just use existing 'exception' key and assign the error, no need for new code.

alikins replied:
Updated to extract any data bundled in the exception into the TaskResult before returning it from the worker.

bcoca replied:
you are still passing the exception object

alikins replied:
Oops, I pushed that to a different branch. Updated now.

bcoca commented on lib/ansible/executor/process/worker.py:
```diff
@@ -118,18 +118,11 @@ def run(self):
             self._rslt_q.put(task_result)
             display.debug("done sending task result")
 
-        except AnsibleConnectionFailure:
```
removing this will now trigger the fallback exception handling, which sets failed=true which is not the same as the previous behavior as unreachable did not affect task status.

bcoca commented on lib/ansible/plugins/connection/ssh.py:
```diff
@@ -552,7 +548,8 @@ def _run(self, cmd, in_data, sudoable=True):
 
         if p.returncode == 255 and in_data:
-            raise AnsibleConnectionFailure('SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh')
+            raise AnsibleConnectionFailure('SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh',
```
again, instead of creating a new way to handle this, just use the existing 'exception' keyword

bcoca commented:
no need for so many changes to add a new facility when the existing ones already can handle this, lets keep it simple and not have 20 paths for everything.

alikins commented:
> no need for so many changes to add a new facility when the existing ones already can handle this, lets keep it simple and not have 20 paths for everything.

Which existing facility are you referring to?

bcoca commented:
passing an 'exception' key in the results

alikins commented:
> passing an 'exception' key in the results

task_executor._execute  (via .run() in the worker) catches the exception from ssh.py, and returns a task result. Adding an exception key there doesn't seem to change anything since strategy/__init__.py _process_pending_results() doesn't transfer the results info for 'unreachable' results, only 'failed'' results.

Even following the existing fail result use of the 'exception' key, the value is just a string dump of the traceback. In this scenario, the exception is expected and handled, so a traceback is misleading. It also doesn't include the stderr info at all which is the useful info. Setting the exception key to the exception msg is slightly better although having stderr as an attribute would be better for display purposes. Doesn't really matter because 'exception' in unreachable results aren't shown.

The 'v2_runner_on_unreachable' callback does get that info, and that would be the way I would prefer to do this (earlier versions of this branch did that but it was veto'ed).

bcoca commented:
I'm not sure that part would be vetoed, i'm fine with adding displaying to that method in the default b callback.

alikins commented:
More varieties:

https://github.com/ansible/ansible/compare/devel...alikins:minimal_better_ssh_failures_callback_data?expand=1  
This is the previous version of this patch, that adds a generic 'data' attribute to all AnsibleErrors, and populates it from task_executor, then the default.py callback can extra the 'stderr' info in the data dict and display it correctly.

https://github.com/ansible/ansible/compare/devel...alikins:minimal_better_ssh_failures_stderr_special_case?expand=1
This version adds only a 'stderr' key to the task executor result, for the special case of AnsibleConnectionFailure. The 'stderr' key will be displayed in normal and verbose modes
as part of the task result data. However, since it will be as  a json string in the default output,
embedded newlines and the like become \n and somewhat hard to read for multiline.

https://github.com/ansible/ansible/compare/devel...alikins:minimal_better_ssh_failures_stderr_special_case_unreach_callback_exc?expand=1
This version adds the stderr and exception key to the task result, and updates default.py callback to treat exceptions in v2_runner_on_unreachable the same as v2_runner_on_fail. The traceback shown in this case for AnsibleConnectionFailure is kind of misleading however. The caveats about the embedded newlines from minimal_better_ssh_failures_stderr_special_case apply here as well.

alikins commented:
Sample output from the other versions:
For this branch (minimal_better_ssh_failures at 4b41967939fba9d6eb4a8884ff

```
[fedora-23:ansible (minimal_better_ssh_failures %)]$ ansible-playbook -i hosts ping_fail.yml 

PLAY [cantconnect,localhost] ***************************************************

TASK [ping] ********************************************************************
ok: [localhost]
fatal: [cantconnect]: UNREACHABLE! => {"changed": false, "msg": "('Failed to connect to the host via ssh.',)\nstderr:\nssh: connect to host 192.168.1.253 port 22: No route to host\r\n", "unreachable": true}

PLAY [cantconnect] *************************************************************
    to retry, use: --limit @/home/adrian/src/ansible/ping_fail.retry

PLAY RECAP *********************************************************************
cantconnect                : ok=0    changed=0    unreachable=1    failed=0   
localhost                  : ok=1    changed=0    unreachable=0    failed=0   

[fedora-23:ansible (minimal_better_ssh_failures %)]$ ansible-playbook -vvv -i hosts ping_fail.yml 
Using /home/adrian/.ansible.cfg as config file

PLAYBOOK: ping_fail.yml ********************************************************
2 plays in ping_fail.yml

PLAY [cantconnect,localhost] ***************************************************

TASK [ping] ********************************************************************
task path: /home/adrian/src/ansible/ping_fail.yml:6
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/system/ping.py
<192.168.1.253> ESTABLISH SSH CONNECTION FOR USER: adrian
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/system/ping.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1475159586.37-50424249666533 `" && echo ansible-tmp-1475159586.37-50424249666533="` echo $HOME/.ansible/tmp/ansible-tmp-1475159586.37-50424249666533 `" ) && sleep 0'
<192.168.1.253> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=adrian -o ConnectTimeout=10 -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r 192.168.1.253 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1475159586.37-163369419432619 `" && echo ansible-tmp-1475159586.37-163369419432619="` echo $HOME/.ansible/tmp/ansible-tmp-1475159586.37-163369419432619 `" ) && sleep 0'"'"''
<localhost> PUT /tmp/tmpYnOFLm TO /home/adrian/.ansible/tmp/ansible-tmp-1475159586.37-50424249666533/ping.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1475159586.37-50424249666533/ /home/adrian/.ansible/tmp/ansible-tmp-1475159586.37-50424249666533/ping.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1475159586.37-50424249666533/ping.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1475159586.37-50424249666533/" > /dev/null 2>&1 && sleep 0'
ok: [localhost] => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "data": null
        }, 
        "module_name": "ping"
    }, 
    "ping": "pong"
}
fatal: [cantconnect]: UNREACHABLE! => {
    "changed": false, 
    "msg": "('Failed to connect to the host via ssh.',)\nstderr:\nssh: connect to host 192.168.1.253 port 22: No route to host\r\n", 
    "unreachable": true
}

PLAY [cantconnect] *************************************************************
    to retry, use: --limit @/home/adrian/src/ansible/ping_fail.retry

PLAY RECAP *********************************************************************
cantconnect                : ok=0    changed=0    unreachable=1    failed=0   
localhost                  : ok=1    changed=0    unreachable=0    failed=0   
```

For minimal_better_ssh_failures_callback_data:

```
[fedora-23:ansible (minimal_better_ssh_failures_callback_data %)]$ ansible-playbook -i hosts ping_fail.yml 

PLAY [cantconnect,localhost] ***************************************************

TASK [ping] ********************************************************************
ok: [localhost]
fatal: [cantconnect]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}
fatal: [cantconnect]: STDERR => 
ssh: connect to host 192.168.1.253 port 22: No route to host


PLAY [cantconnect] *************************************************************
    to retry, use: --limit @/home/adrian/src/ansible/ping_fail.retry

PLAY RECAP *********************************************************************
cantconnect                : ok=0    changed=0    unreachable=1    failed=0   
localhost                  : ok=1    changed=0    unreachable=0    failed=0   

[fedora-23:ansible (minimal_better_ssh_failures_callback_data %)]$ ansible-playbook -vvv -i hosts ping_fail.yml 
Using /home/adrian/.ansible.cfg as config file

PLAYBOOK: ping_fail.yml ********************************************************
2 plays in ping_fail.yml

PLAY [cantconnect,localhost] ***************************************************

TASK [ping] ********************************************************************
task path: /home/adrian/src/ansible/ping_fail.yml:6
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/system/ping.py
<192.168.1.253> ESTABLISH SSH CONNECTION FOR USER: adrian
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/system/ping.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1475159330.59-208955447526426 `" && echo ansible-tmp-1475159330.59-208955447526426="` echo $HOME/.ansible/tmp/ansible-tmp-1475159330.59-208955447526426 `" ) && sleep 0'
<192.168.1.253> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=adrian -o ConnectTimeout=10 -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r 192.168.1.253 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1475159330.59-128481110596099 `" && echo ansible-tmp-1475159330.59-128481110596099="` echo $HOME/.ansible/tmp/ansible-tmp-1475159330.59-128481110596099 `" ) && sleep 0'"'"''
<localhost> PUT /tmp/tmpRuz5Ms TO /home/adrian/.ansible/tmp/ansible-tmp-1475159330.59-208955447526426/ping.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1475159330.59-208955447526426/ /home/adrian/.ansible/tmp/ansible-tmp-1475159330.59-208955447526426/ping.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1475159330.59-208955447526426/ping.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1475159330.59-208955447526426/" > /dev/null 2>&1 && sleep 0'
ok: [localhost] => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "data": null
        }, 
        "module_name": "ping"
    }, 
    "ping": "pong"
}
fatal: [cantconnect]: UNREACHABLE! => {
    "changed": false, 
    "msg": "Failed to connect to the host via ssh.", 
    "unreachable": true
}
fatal: [cantconnect]: STDERR => 
ssh: connect to host 192.168.1.253 port 22: No route to host


PLAY [cantconnect] *************************************************************
    to retry, use: --limit @/home/adrian/src/ansible/ping_fail.retry

PLAY RECAP *********************************************************************
cantconnect                : ok=0    changed=0    unreachable=1    failed=0   
localhost                  : ok=1    changed=0    unreachable=0    failed=0   
```

For minimal_better_ssh_failures_stderr_special_case:

```
[fedora-23:ansible (minimal_better_ssh_failures_stderr_special_case %)]$ ansible-playbook -i hosts ping_fail.yml 

PLAY [cantconnect,localhost] ***************************************************

TASK [ping] ********************************************************************
ok: [localhost]
fatal: [cantconnect]: UNREACHABLE! => {"changed": false, "msg": "('Failed to connect to the host via ssh.',)", "stderr": "ssh: connect to host 192.168.1.253 port 22: No route to host\r\n", "unreachable": true}

PLAY [cantconnect] *************************************************************
    to retry, use: --limit @/home/adrian/src/ansible/ping_fail.retry

PLAY RECAP *********************************************************************
cantconnect                : ok=0    changed=0    unreachable=1    failed=0   
localhost                  : ok=1    changed=0    unreachable=0    failed=0   

[fedora-23:ansible (minimal_better_ssh_failures_stderr_special_case %)]$ ansible-playbook -vvv -i hosts ping_fail.yml 
Using /home/adrian/.ansible.cfg as config file

PLAYBOOK: ping_fail.yml ********************************************************
2 plays in ping_fail.yml

PLAY [cantconnect,localhost] ***************************************************

TASK [ping] ********************************************************************
task path: /home/adrian/src/ansible/ping_fail.yml:6
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/system/ping.py
<192.168.1.253> ESTABLISH SSH CONNECTION FOR USER: adrian
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/system/ping.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1475159388.69-170230821072158 `" && echo ansible-tmp-1475159388.69-170230821072158="` echo $HOME/.ansible/tmp/ansible-tmp-1475159388.69-170230821072158 `" ) && sleep 0'
<192.168.1.253> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=adrian -o ConnectTimeout=10 -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r 192.168.1.253 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1475159388.69-111512491736865 `" && echo ansible-tmp-1475159388.69-111512491736865="` echo $HOME/.ansible/tmp/ansible-tmp-1475159388.69-111512491736865 `" ) && sleep 0'"'"''
<localhost> PUT /tmp/tmpegVxOB TO /home/adrian/.ansible/tmp/ansible-tmp-1475159388.69-170230821072158/ping.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1475159388.69-170230821072158/ /home/adrian/.ansible/tmp/ansible-tmp-1475159388.69-170230821072158/ping.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1475159388.69-170230821072158/ping.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1475159388.69-170230821072158/" > /dev/null 2>&1 && sleep 0'
ok: [localhost] => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "data": null
        }, 
        "module_name": "ping"
    }, 
    "ping": "pong"
}
fatal: [cantconnect]: UNREACHABLE! => {
    "changed": false, 
    "msg": "('Failed to connect to the host via ssh.',)", 
    "stderr": "ssh: connect to host 192.168.1.253 port 22: No route to host\r\n", 
    "unreachable": true
}

PLAY [cantconnect] *************************************************************
    to retry, use: --limit @/home/adrian/src/ansible/ping_fail.retry

PLAY RECAP *********************************************************************
cantconnect                : ok=0    changed=0    unreachable=1    failed=0   
localhost                  : ok=1    changed=0    unreachable=0    failed=0   
```

For minimal_better_ssh_failures_stderr_special_case_unreach_callback_exc: 

```
[fedora-23:ansible (minimal_better_ssh_failures_stderr_special_case_unreach_callback_exc %)]$ ansible-playbook -i hosts ping_fail.yml 

PLAY [cantconnect,localhost] ***************************************************

TASK [ping] ********************************************************************
ok: [localhost]
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AnsibleSshConnectionFailure: ('Failed to connect to the host via ssh.',)
fatal: [cantconnect]: UNREACHABLE! => {"changed": false, "msg": "('Failed to connect to the host via ssh.',)", "stderr": "ssh: connect to host 192.168.1.253 port 22: No route to host\r\n", "unreachable": true}

PLAY [cantconnect] *************************************************************
    to retry, use: --limit @/home/adrian/src/ansible/ping_fail.retry

PLAY RECAP *********************************************************************
cantconnect                : ok=0    changed=0    unreachable=1    failed=0   
localhost                  : ok=1    changed=0    unreachable=0    failed=0   

[fedora-23:ansible (minimal_better_ssh_failures_stderr_special_case_unreach_callback_exc %)]$ ansible-playbook -vvv -i hosts ping_fail.yml 
Using /home/adrian/.ansible.cfg as config file

PLAYBOOK: ping_fail.yml ********************************************************
2 plays in ping_fail.yml

PLAY [cantconnect,localhost] ***************************************************

TASK [ping] ********************************************************************
task path: /home/adrian/src/ansible/ping_fail.yml:6
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/system/ping.py
<192.168.1.253> ESTABLISH SSH CONNECTION FOR USER: adrian
Using module file /home/adrian/src/ansible/lib/ansible/modules/core/system/ping.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1475159482.3-30333130798778 `" && echo ansible-tmp-1475159482.3-30333130798778="` echo $HOME/.ansible/tmp/ansible-tmp-1475159482.3-30333130798778 `" ) && sleep 0'
<192.168.1.253> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=adrian -o ConnectTimeout=10 -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r 192.168.1.253 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1475159482.3-144762396335971 `" && echo ansible-tmp-1475159482.3-144762396335971="` echo $HOME/.ansible/tmp/ansible-tmp-1475159482.3-144762396335971 `" ) && sleep 0'"'"''
<localhost> PUT /tmp/tmpXgb8Aw TO /home/adrian/.ansible/tmp/ansible-tmp-1475159482.3-30333130798778/ping.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1475159482.3-30333130798778/ /home/adrian/.ansible/tmp/ansible-tmp-1475159482.3-30333130798778/ping.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1475159482.3-30333130798778/ping.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1475159482.3-30333130798778/" > /dev/null 2>&1 && sleep 0'
ok: [localhost] => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "data": null
        }, 
        "module_name": "ping"
    }, 
    "ping": "pong"
}
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/home/adrian/src/ansible/lib/ansible/executor/task_executor.py", line 488, in _execute
    result = self._handler.run(task_vars=variables)
  File "/home/adrian/src/ansible/lib/ansible/plugins/action/normal.py", line 33, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars))
  File "/home/adrian/src/ansible/lib/ansible/plugins/action/__init__.py", line 616, in _execute_module
    tmp = self._make_tmp_path(remote_user)
  File "/home/adrian/src/ansible/lib/ansible/plugins/action/__init__.py", line 220, in _make_tmp_path
    result = self._low_level_execute_command(cmd, sudoable=False)
  File "/home/adrian/src/ansible/lib/ansible/plugins/action/__init__.py", line 766, in _low_level_execute_command
    rc, stdout, stderr = self._connection.exec_command(cmd, in_data=in_data, sudoable=sudoable)
  File "/home/adrian/src/ansible/lib/ansible/plugins/connection/ssh.py", line 607, in exec_command
    stderr=return_tuple[2])
AnsibleSshConnectionFailure: ('Failed to connect to the host via ssh.',)

fatal: [cantconnect]: UNREACHABLE! => {
    "changed": false, 
    "msg": "('Failed to connect to the host via ssh.',)", 
    "stderr": "ssh: connect to host 192.168.1.253 port 22: No route to host\r\n", 
    "unreachable": true
}

PLAY [cantconnect] *************************************************************
    to retry, use: --limit @/home/adrian/src/ansible/ping_fail.retry

PLAY RECAP *********************************************************************
cantconnect                : ok=0    changed=0    unreachable=1    failed=0   
localhost                  : ok=1    changed=0    unreachable=0    failed=0   
```

alikins committed:
Remove AnsibleSshConnectionFailure

Use generic AnsibleConnectionFailure instead.

alikins committed:
Add ssh stderr to AnsibleConnectionFailure msg

Remove use of AnsibleSshConnectionFailure, and instead of
passing the stderr as an attribute on the exception, concatinate
it into the AnsibleConnectionFailure msg.

alikins committed:
Fix unusued imports and unit test.

bcoca commented:
closed via https://github.com/ansible/ansible/commit/9ec69ee7eaf6386e0931a7826cc413f128554d89
